File: c++/jan/25_23/longest_palindromic_substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        if (s.empty()) return ""; // If the string is empty
        int start = 0, maxLength = 0;

        // Helper function to expand around the center
        auto expandAroundCenter = [&](int left, int right) {
            while (left >= 0 && right < s.length() && s[left] == s[right]) {
                left--;
                right++;
            }
            return right - left - 1; // Length of the palindrome
        };

        for (int i = 0; i < s.length(); i++) {
            // Expand around a single character (odd-length palindromes)
            int len1 = expandAroundCenter(i, i);
            // Expand around two characters (even-length palindromes)
            int len2 = expandAroundCenter(i, i + 1);
            // Find the maximum length palindrome for this center
            int len = max(len1, len2);
            if (len > maxLength) {
                maxLength = len;
                start = i - (len - 1) / 2; // Calculate starting index of palindrome
            }
        }

        // Return the longest palindromic substring
        return s.substr(start, maxLength); 
    }
};

File: python/jan/25_23/longest_palindromic_substring.py
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        if not s:
            return ""
        
        start, max_length = 0, 0

        # Helper function to expand around the center
        def expandAroundCenter(left, right):
            while left >= 0 and right < len(s) and s[left] == s[right]:
                left -= 1
                right += 1
            return right - left - 1  # Length of the palindrome

        for i in range(len(s)):
            # Odd-length palindromes (center at one character)
            len1 = expandAroundCenter(i, i)
            # Even-length palindromes (center between two characters)
            len2 = expandAroundCenter(i, i + 1)
            # Find the maximum length for the current center
            curr_length = max(len1, len2)
            if curr_length > max_length:
                max_length = curr_length
                start = i - (curr_length - 1) // 2  # Update start index of the palindrome

        # Return the longest palindromic substring
        return s[start:start + max_length]
